tests/line.c

```diff
@@ -150,70 +150,274 @@ void svpng(FILE *fp, unsigned w, unsigned h, const uint8_t *img, bool alpha)
+
+    tmp = q_add(q_mul(dx, dx), q_mul(dy, dy));
+    q_fmt res = q_add(sqrtq(tmp), -r);
+    return res;
 }
 
 /* Render shapes into the buffer individually with alpha blending. */
-void alphablend(int x, int y, float alpha, float r, float g, float b)
+void alphablend(int x, int y, q_fmt alpha, q_fmt r, q_fmt g, q_fmt b)
 {
     uint8_t *p = img + (y * W + x) * 3;
-    p[0] = (uint8_t) (p[0] * (1 - alpha) + r * alpha * 255);
-    p[1] = (uint8_t) (p[1] * (1 - alpha) + g * alpha * 255);
-    p[2] = (uint8_t) (p[2] * (1 - alpha) + b * alpha * 255);
+    p[0] = (uint8_t) Q2I(
+        q_add(p[0] * q_add((1 << Q), -alpha), q_mul(r, alpha) * 255));
+    p[1] = (uint8_t) Q2I(
+        q_add(p[1] * q_add((1 << Q), -alpha), q_mul(g, alpha) * 255));
+    p[2] = (uint8_t) Q2I(
+        q_add(p[2] * q_add((1 << Q), -alpha), q_mul(b, alpha) * 255));
 }
 
 /* Use AABB of capsule to reduce the number of samples. */
-void lineSDFAABB(float ax, float ay, float bx, float by, float r)
+void lineSDFAABB(q_fmt ax, q_fmt ay, q_fmt bx, q_fmt by, q_fmt r)
 {
-    int x0 = (int) floorf(fminf(ax, bx) - r);
-    int x1 = (int) ceilf(fmaxf(ax, bx) + r);
-    int y0 = (int) floorf(fminf(ay, by) - r);
-    int y1 = (int) ceilf(fmaxf(ay, by) + r);
+    int x0 = Q2I(floorq(q_add(min(ax, bx), -r)));
+    int x1 = Q2I(ceilq(q_add(max(ax, bx), r)));
+    int y0 = Q2I(floorq(q_add(min(ay, by), -r)));
+    int y1 = Q2I(ceilq(q_add(max(ay, by), r)));
     for (int y = y0; y <= y1; y++) {
         for (int x = x0; x <= x1; x++)
-            alphablend(
-                x, y,
-                fmaxf(fminf(0.5f - capsuleSDF(x, y, ax, ay, bx, by, r), 1.0f),
-                      0.0f),
-                0.0f, 0.0f, 0.0f);
+            alphablend(x, y,
+                       max(min((1 << (Q - 1)) - capsuleSDF(I2Q(x), I2Q(y), ax,
+                                                           ay, bx, by, r),
+                               I2Q(1)),
+                           0),
+                       0, 0, 0);
     }
 }
 
 int main()
 {
     memset(img, 255, sizeof(img));
-    float cx = W * 0.5f, cy = H * 0.5f;
+    q_fmt cx = W * (1 << (Q - 1)), cy = H * (1 << (Q - 1));
+    q_fmt ct, st;
     for (int j = 0; j < 5; j++) {
-        float r1 = fminf(W, H) * (j + 0.5f) * 0.085f;
-        float r2 = fminf(W, H) * (j + 1.5f) * 0.085f;
-        float t = j * PI / 64.0f, r = (j + 1) * 0.5f;
-        for (int i = 1; i <= 64; i++, t += 2.0f * PI / 64.0f) {
-            float ct = cosf(t), st = sinf(t);
-            lineSDFAABB(cx + r1 * ct, cy - r1 * st, cx + r2 * ct, cy - r2 * st,
-                        r);
+        q_fmt r1 = min(W, H) * q_mul((I2Q(j) + (1 << (Q - 1))), f2Q(0.085f));
+        q_fmt r2 = min(W, H) * q_mul((I2Q(j) + (3 << (Q - 1))), f2Q(0.085f));
+        q_fmt t = j * q_div(PI_Q, I2Q(64));
+        q_fmt r = (j + 1) * (1 << (Q - 1));
+        for (int i = 1; i <= 64;
+             i++, t = q_add(t, q_mul(I2Q(2), q_div(PI_Q, I2Q(64))))) {
```

Comment:
Move the statement `t = q_add(t, q_mul(I2Q(2), q_div(PI_Q, I2Q(64))))` right after `lineSDFAABB`.